### PR TITLE
Prevent Index & Delete request primaryTerm getter/setter, setShardId setter

### PR DESCRIPTION
**Pull request by dakrone**
`IndexRequest` and `DeleteRequest` inherit from `ReplicationRequest`, which
allows getting and setting the primary term for a particular operation. However,
in this case, either request will be wrapped into a `BulkItemRequest` wrapper,
which is where the primary term and shardId should actually be set. We
(dangerously) use the request object to store some internal state about the
request.

This PR makes that internal state for just the Index and Delete request throw an
exception, so that it is clear that it should never be set or retrieved from the
inner object. Originally we intended to have separate objects
`LegacyIndexRequest` and `LegacyDeleteRequest` for removing that functionality,
however the wrappers required for client compatibility would make the complexity
trade-off not worth it. We plan to revisit the removal of this internal state
after the transport client has been removed.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/delete/DeleteRequest.java [hunk 1]
```diff
@@ -221,4 +222,24 @@ public void writeTo(StreamOutput out) throws IOException {
     public String toString() {
         return "delete {[" + index + "][" + type + "][" + id + "]}";
     }
+
+    // Override these methods from ReplicationAction, these are things where we are storing our state in the request object (which we really
+    // shouldn't do). Once the transport client goes away we can move away from making these available, but in the meantime these are
+    // dangerous to set or use because the DeleteRequest object will always be wrapped in a bulk request envelope, which is where these
+    // *should* be set.
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/delete/DeleteRequest.java]: Please promote these to legitimate Javadocs on each method so they show up in IDEs, etc.

File: core/src/main/java/org/elasticsearch/action/delete/DeleteRequest.java [hunk 2]
```diff
@@ -221,4 +222,24 @@ public void writeTo(StreamOutput out) throws IOException {
...
+
+    @Override
+    public long primaryTerm() {
+        throw new UnsupportedOperationException("primary should never be set on DeleteRequest");
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/delete/DeleteRequest.java]: Missing "term"?

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left some comments. Also can you provide a PR description/commit message that motivates what is going on here.

**Comment by dakrone** [comment]
@jasontedor sure, updated the PR with a description and a commit.

File: core/src/main/java/org/elasticsearch/action/delete/DeleteRequest.java [hunk 3]
```diff
@@ -221,4 +222,34 @@ public void writeTo(StreamOutput out) throws IOException {
...
+    }
+
+    /**
+     * Override this method from ReplicationAction, this is where we are storing our state in the request object (which we really shouldn't
+     * do). Once the transport client goes away we can move away from making this available, but in the meantime this is dangerous to set or
+     * use because the DeleteRequest object will always be wrapped in a bulk request envelope, which is where this *should* be set.
+     */
+    @Override
+    public void primaryTerm(long term) {
+        throw new UnsupportedOperationException("primary should never be set on DeleteRequest");
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/delete/DeleteRequest.java]: Sorry, I wasn't explicit, but you're missing "term" everywhere except for the ones that you just fixed.

**Review by jasontedor — CHANGES_REQUESTED** [review]
You're missing "term".

**Review by jasontedor — APPROVED** [review]
LGTM.

**Pull request merged**